### PR TITLE
BUG: eps param no effect fixed

### LINKED ISSUE #16600: BUG: `KDTree`'s optional argument `eps` seems to have no effect
### Describe your issue.

When calling `query_pairs()` or `query_ball_tree()` from the `KDTree`-object, the `eps`-argument seems to have no effect (this might hold for other `query_`-methods as well). This becomes apparent when the distance between nodes of a grid is an irrational number (e.g. `np.sqrt(2)`) and the search radius (i.e. `r`) is as well. Due to rounding errors introduced by working with irrational numbers, the calculated distance might become slightly larger than the radius, which excludes the node(s) from the "connecting". It seems that the `eps`-argument is introduced (partly) because of this, but it is not working; see the MWE below.

### Reproducing Code Example

```python
import numpy as np
from scipy.spatial import KDTree

# create an evenly distributed grid with a spacing of np.sqrt(2)
spacing = np.sqrt(2)
x_range = np.linspace(0, 3 * spacing, 4)
y_range = np.linspace(0, 3 * spacing, 4)
xy_array = [(xi, yi) for xi in x_range for yi in y_range]

# initiate the KDTree
tree = KDTree(xy_array)

# search for the internal combinations of points
# the correct answer is 24 pairs, i.e. `len(pairs) = 24`
pairs = tree.query_pairs(r=spacing)
print(len(pairs))
# result: 16

# using `eps` to overcome the rounding errors
pairs = tree.query_pairs(r=spacing, eps=.1)
print(len(pairs))
# result: 16

# the above should result in the same
pairs = tree.query_pairs(r=1.01 * spacing)
print(len(pairs))
# result: 24
```

### Error message

```shell
-
```

### SciPy/NumPy/Python version information

1.7.3 1.19.4 sys.version_info(major=3, minor=7, micro=0, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by njj67229**
#### Reference issue
Closes #16600

#### What does this implement/fix?
Adds eps to be considered in kdtree's `query_pairs()` function

#### Additional information
Would you like me to implement a test in `scipy/spatial/tests/test_kdtree.py` that uses code similar to the example in issue #16600?

### CODE REVIEW

**Review by tupui — COMMENTED** [review]
Thank you for the PR @njj67229. A test demonstrating that the parameter is being used would indeed be interesting to have.

**Review by EngNashar — APPROVED** [review]

File: scipy/spatial/tests/test_kdtree.py [hunk 1]
```diff
@@ -782,6 +782,21 @@ def test_kdtree_query_pairs(kdtree_type):
     assert_array_equal(l0, l2)
 
 
+def test_query_pairs_eps(kdtree_type):
+        spacing = np.sqrt(2) 
+        # irrational spacing to have potential rounding errors
```
Comment (tupui) [inline scipy/spatial/tests/test_kdtree.py:787]: The CI is failing due to over indentation here.

File: scipy/spatial/tests/test_kdtree.py [hunk 2]
```diff
@@ -782,6 +782,20 @@ def test_kdtree_query_pairs(kdtree_type):
     assert_array_equal(l0, l2)
 
 
+def test_query_pairs_eps(kdtree_type):
+    spacing = np.sqrt(2)
```
Comment (tupui) [inline scipy/spatial/tests/test_kdtree.py:786]:
```suggestion
    spacing = np.sqrt(2)
```

**Review by tupui — APPROVED** [review]
LGTM I can verify the numbers, thanks @njj67229

**Comment by tylerjereddy** [comment]
Seem to cherry-pick cleanly to the maintenance branch so I'll add a tentative backport label given this was considered a bug.

**Pull request merged**